src/pages/filter/common/dashboard/components/body/dashboard-body.tsx

```diff
@@ -1,155 +1,40 @@
-        setSpareTabs(prev => [selectedGroup!, ...prev])
-        setMainTabsLayout(prev => prev.filter(group => group.i !== groupName))
-      } else {
-        setSpareTabs(prev => prev.filter((_, index) => index !== groupIndex))
-        setMainTabs(prev => [...prev, selectedGroup!])
-        const newTabLayout = getNewTabLayout(selectedGroup!, mainTabsLayout)
-        setMainTabsLayout(newTabLayout)
-      }
-    }
-
-    const changeTabHeight = ({ index, id, isOpen }: IChangeHeightProps) => {
-      const newLayout = getLayoutOnTabHeightChange({
-        index,
-        id,
-        isOpen,
-        mainTabsLayout,
-      })
-      setMainTabsLayout(newLayout)
-    }
-
-    const changeSubTabHeight = ({ index, id, isOpen }: IChangeHeightProps) => {
-      const newLayout = getLayoutOnSubTabHeightChange({
-        index,
-        id,
-        isOpen,
-        mainTabsLayout,
-      })
-      setMainTabsLayout(newLayout)
-    }
-
-    const makeTabFavorite = (
-      groupType: DashboardGroupTypes,
-      groupName: string,
-      groupIndex: number,
-    ) => {
-      if (groupType === DashboardGroupTypes.Main) {
-        const clonedTabs = cloneDeep(mainTabs)
-
-        clonedTabs.forEach((item, index) => {
-          if (item.name === groupName) {
-            clonedTabs[index].isFavorite = !clonedTabs[index].isFavorite
-          }
-        })
-
-        const sortedTabs = getSortedTabs(clonedTabs)
-        setMainTabs(sortedTabs)
-        setMainTabsLayout(getUpdatedLayoutLayout(sortedTabs, mainTabsLayout))
-      } else {
-        const selectedGroup = groups.find(group => group.name === groupName)
-        setSpareTabs(prev => prev.filter((_, index) => index !== groupIndex))
-
-        const newMainTabs = [...mainTabs, selectedGroup!]
-        const newLayout = getNewTabLayout(selectedGroup!, mainTabsLayout)
-
-        newMainTabs.forEach((item, index) => {
-          if (item.name === groupName) {
-            newMainTabs[index].isFavorite = !newMainTabs[index].isFavorite
-          }
-        })
-
-        const sortedTabs = getSortedTabs(newMainTabs)
-        setMainTabs(sortedTabs)
-        setMainTabsLayout(getUpdatedLayoutLayout(sortedTabs, newLayout))
-      }
-    }
-
-    const handleLayoutChange = (layout: Layout[]) => {
-      LocalStoreManager.write(
-        'dashboardLayout',
-        layout,
-        datasetStore.datasetName,
-      )
-
-      setMainTabsLayout(layout)
-    }
-
-    useEffect(() => {
-      LocalStoreManager.write(
-        'dashboardMainTabs',
-        mainTabs,
-        datasetStore.datasetName,
-      )
-      LocalStoreManager.write(
-        'dashboardLayout',
-        mainTabsLayout,
-        datasetStore.datasetName,
-      )
-    }, [mainTabs, mainTabsLayout])
+  ({ className }: IDashboardBodyProps): ReactElement => {
+    const {
+      changeTabsHeight,
+      mainTabs,
+      mainTabsLayout,
+      showInCharts,
+      layoutChange,
+    } = dashboardStore
+
+    useLayoutEffect(() => {
```

Comment:
What if we move this logic into the dashboard.store reaction (with the same timeout) ?

Reply:
It doesn't work there, if we move it to store it will give us same layout as we have already, because browser need time to redraw. if we start to increment timeout it will be inaccurate, because some time of re rendering maybe different. If we remove here timeout it will also give us wrong layout if we have charts, case they need more time to render